daml-lf/spec/daml-lf-1.rst

```diff
@@ -2569,6 +2575,150 @@ Map functions
 
   [*Available in versions >= 1.3*]
 
+
+Generic Map functions
+~~~~~~~~~~~~~~~~~~~~~
+
+* ``GENMAP_EMPTY : ∀ α. ∀ β. 'GenMap' α β``
+
+  Returns the empty generic map.
+
+  [*Available in versions >= 1.dev*]
+
+* ``GENMAP_INSERT : ∀ α. ∀ β.  α → β → 'GenMap' α β → 'GenMap' α β``
+
+  Inserts a new key and value in the map. If the key is already
+  present in the map, the associated value is replaced with the
+  supplied value.
```

Comment:
We should specify how this affects the "order in which keys were inserted" which is relevant for the conversion to lists, do we first-inserted-order or last-inserted-order? If we go with the former, the latter can be _efficiently_ expressed in terms of the former (and `GENMAP_DELETE`). I don't see how this would be possible _efficiently_ the other way around.

Reply:
I think first inserted order makes more sense. It's certainly what I meant in the spec. It implies that if you insert the same (key, value) pair in a map that is already there, it will not modify the map, i.e. inserts are idempotent. That's nice.

Reply:
I don't know what you mean _exactly_ by "inserts are idempotent". I assume you mean
```
insert k v . insert k v  ==  insert k v
```
for all `k` and `v`. This is true for last-inserted order as well. I think what is more compelling about first-inserted order is that we get
```
insert k (lookup k m) m  ==  m
```
for all `m` and keys `k` of `m`.

We can also define
```
insertLast k v = insertFirst k v . delete k
```
Thus, insert-first order is at least as "powerful" as insert-last order.

Reply:
Ah, you're right. My sentence was better before the `i.e.`.